servers/rendering/renderer_rd/renderer_canvas_render_rd.cpp

```diff
@@ -1627,6 +1635,38 @@ RendererCanvasRenderRD::RendererCanvasRenderRD() {
 
 		default_samplers.default_filter = RS::CANVAS_ITEM_TEXTURE_FILTER_LINEAR;
 		default_samplers.default_repeat = RS::CANVAS_ITEM_TEXTURE_REPEAT_DISABLED;
+		Vector<RID> samplers = material_storage->samplers_rd_get_default().get_rids();
+		for (uint32_t i = 0; i < DEFAULT_MATERIAL_SAMPLER_COUNT; i++) {
+			default_material_samplers[i] = samplers[i];
+		}
+	}
+
+	uint32_t push_pad_size = 0;
+
+	{
+		// determine max number of texture slots we can use, which is a function of two driver limits:
+		//
+		// - LIMIT_MAX_TEXTURES_PER_SHADER_STAGE, which is the maximum number of texture slots per stage
+		// - LIMIT_MAX_PUSH_CONSTANT_SIZE, which is the maximum size (in bytes) of a push constant buffer
+
+		uint32_t available_texture_slots = MIN(RD::get_singleton()->limit_get(RD::LIMIT_MAX_TEXTURES_PER_SHADER_STAGE) - TEXTURE_SLOTS_USED, BATCH_MAX_TEXTURES);
+
+		// base_pc_size is the fixed-size of the data in the push constant, so we can
+		// determine how much is available for the specular_shininess data
+		uint32_t base_pc_size = offsetof(PushConstant, specular_shininess);
+		// based on the maximum size of the push constant from the driver, how many bytes are available
+		// for the specular_shininess data
+		uint32_t available_pc_bytes = RD::get_singleton()->limit_get(RD::LIMIT_MAX_PUSH_CONSTANT_SIZE) - base_pc_size;
```

Comment:
Prior to the 2D batching work that was recently merged, the 2D renderer was using a 128 byte push constant for each canvas instance. How would that have affected mobile devices?

Sounds like I should use another uniform buffer for the specular shininess for each batch?

Reply:
> Prior to the 2D batching work that was recently merged, the 2D renderer was using a 128 byte push constant for each draw call to draw a single Canvas Item. How would that have affected mobile devices?

It was bad. And we had plans to change it once we implemented batching. To be clear, the performance impact may not be that significant anymore (especially since we can now draw an entire scene in a handful of draw calls). Additionally newer mobile devices may properly support large push constants (I see many devices advertise support for 256 bytes now). 

Internally, older mobile devices are doing a uniform buffer update call every time you set a push constant and then placing their own index in the 16 byte push constant. This creates an extra level of indirection and results in a lot of small buffer updates. Which is clearly much worse than us maintaining our own uniform buffer that gets updated once per frame. 

> Sounds like I should use another uniform buffer for the specular shininess for each batch?

You could. 

I also tested locally with setting a hard limit of 16 bytes and it works as well as regular batching. So that's an option too

Reply:
How do I query this hard limit for mobile devices specifically? If we can query for Vulkan + Mobile, we can safely set that hard limit, without penalising other devices.

If I set the 16-byte limit we can only batch 3 unique textures with the current implementation, as the shininess value is paired with the `texture_array` uniform, so they need an equal number of slots. 

Whilst typing and thinking about this, I have figured out a way to support a single, per-frame uniform, for shininess. 😄 

----

I am _very_ much looking forward to UMA support, as that would be a huge win here. I'd be able to utilise that for the instance buffers for a pretty significant win on UMA devices. Without UMA:

- write instance data into a CPU buffer
- once that is full, we then call `buffer_update`, which gets copied to a staging buffer and that gets "uploaded" to the GPU (into chunks), where upload is not necessary for UMA
- that gets scheduled to be blitted in to the instance buffer that is bound to the draw call.

With UMA we would eliminate all the copies and blits and write directly to the instance buffer that is bound to the draw call.